### ISSUE TITLE
Large memory is used by volcano-scheduler

### ISSUE THREAD

**Issue by jazka** [volcano-sh/volcano#1947]
**What happened**:
The volcano-scheduler uses large memory and the function of schedule seems to be unusual. The pod of job will always be pending. But there is no any error logs in pod. After restart it, everything is okay. 

The image link of the result of  "kubectl top pods -n volcano-system":
https://l4x826wg3c.feishu.cn/file/boxcnLnrbgq6CmvjAQWfOlWX9gc

**What you expected to happen**:
The schduler works well. Or know how to check what happened, then I could restart it when monitor some events.

**How to reproduce it (as minimally and precisely as possible)**:
Not sure how to reproduce it, but it happend many times after running some days.

**Anything else we need to know?**:
1. The ETCD in our cluster uses normal Disk, not SSD. 
2. The Node volcano pods are deployed could be scheduled by other computer task like pytorchjob, and there is no Requests to volcano.

**Environment**:
- Volcano Version: https://github.com/volcano-sh/volcano/commit/1b96bdf4de821e1e4af2b4c056f67be7559a880d
- Kubernetes version (use `kubectl version`): 1.22.2
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by hwdef** [comment]
Please provide more information such as scheduler logs

**Comment by jazka** [comment]
> Please provide more information such as scheduler logs

The images below are the some logs screenshot (I can not parse all the logs here, sorry for that), but there is no other different contents with these logs.

https://l4x826wg3c.feishu.cn/file/boxcnFFuid08JWiylxu0NAw5CZc
https://l4x826wg3c.feishu.cn/file/boxcnLnyvqitaA6bAshUriNUEMc
https://l4x826wg3c.feishu.cn/file/boxcnQmvPxfjGiNv6BglGmDH7Za
https://l4x826wg3c.feishu.cn/file/boxcnYDoTEobLvJZ0bPHx4vMpFb

**Comment by k82cn** [comment]
did you check whether the memory of scheduler pod keep increasing?

**Comment by jazka** [comment]
I am not very sure if it keeps increasing. The large memory case was found after scheduler has been running for a few days, I guess the scheduler has some internal errors in these days. My guess is from the logs of the scheduler which has no more  scheduling logs to output.
After the large memory used by scheduler pod was found, I watched it about one hour, it did not keey increasing. Then I restarted it to restore the scheduler functions since all pods used volcano are pending in cluster.

Besides the scheduler did not working again yesterday after it is restarted and running for some hours. I checked the memory used again, it was not very large, abount 1Gi. But I checked the logs and found it may be stuck in function "SchedulerCache => Snapshot". The logs screenshot attached https://l4x826wg3c.feishu.cn/file/boxcnbNoF6JCTdbNEKoX6i0yUpb. 

I review the scheduler code, it should be period to run "Scheduler => runOnce", but it only outputs "Start scheduling ..." and no "End scheduling ..." 
https://github.com/volcano-sh/volcano/blob/f70a8187995f2a4fd3128e37fb9b8d839daf9dff/pkg/scheduler/scheduler.go#L81-L110

The last useful log is "[cache.go:969]: Namespace xxx has weight 1", it comes from function "SchedulerCache => Snapshot". If it works well, we will see "There are <%d> Jobs, <%d> Queues and <%d> Nodes in total for scheduling." in the log, but it is disappeared.
https://github.com/volcano-sh/volcano/blob/f70a8187995f2a4fd3128e37fb9b8d839daf9dff/pkg/scheduler/cache/cache.go#L859-L887


BTW, how to find the source code corresponding to the image. We deployed volcano from https://raw.githubusercontent.com/volcano-sh/volcano/release-1.4/installer/volcano-development.yaml, but the image tag in yaml is "latest", I am not sure what is the source code commit exactly.

**Comment by hwdef** [comment]
you can enter the container, and exec 
```shell
./vc-scheduler --version
```

**Comment by jazka** [comment]
> 

Cool, it is 1b96bdf4de821e1e4af2b4c056f67be7559a880d, is it not the recommended version?

**Comment by hwdef** [comment]
It doesn't look like an official release, somewhere between 1.4 and 1.5.
If it is allowed, is it possible to use the 1.4 mirror test to see if there is still this problem?

**Comment by jazka** [comment]
Sure, I will try it with 1.4 version.

**Comment by william-wang** [comment]
> Besides the scheduler did not working again yesterday after it is restarted and running for some hours. I checked the memory used again, it was not very large, abount 1Gi. But I checked the logs and found it may be stuck in function "SchedulerCache => Snapshot". The logs screenshot attached https://l4x826wg3c.feishu.cn/file/boxcnbNoF6JCTdbNEKoX6i0yUpb.
> 
> I review the scheduler code, it should be period to run "Scheduler => runOnce", but it only outputs "Start scheduling ..." and no "End scheduling ..."
> 
> https://github.com/volcano-sh/volcano/blob/f70a8187995f2a4fd3128e37fb9b8d839daf9dff/pkg/scheduler/scheduler.go#L81-L110
> 
> The last useful log is "[cache.go:969]: Namespace xxx has weight 1", it comes from function "SchedulerCache => Snapshot". If it works well, we will see "There are <%d> Jobs, <%d> Queues and <%d> Nodes in total for scheduling." in the log, but it is disappeared.
> 
Did the message"The priority of job xxx"(Line955 in cache.go) appeared in log? Not sure whether it was stuck in cloneJob.
I am running the same version in my environment and try to reproduce the issue.

**Comment by jazka** [comment]
No, it did not output the log "The priority of job xxx is xxx". Yes, I also guess it was stuck in "SchedulerCache => Snapshot", something like deak lock or sync wait. 

If you want to reporduce this issue, it is better to test with training-operator pytorchjob, since it will create podgroup by call volcano sdk directly. Think the case below:
1. Disable gang-scheduler in training-operator, then it will create pytorchjob, and without podgroup. The pods of pytorchjob will be scheduled by "default-scheduler"
2. Enable gang-scheduler in training-operator, then it will reconcile and call volcano sdk to create podgourp for the running job, even its pods has been running with "schedulerName: default-scheduler"
I am not sure if it is related with volcano stuck. I am also tring to find the way to reproduce this issue.

The related code in training-oprator:
https://github.com/kubeflow/common/blob/432268369798aa796a6f38fddbeae8abc4120e24/pkg/controller.v1/common/job_controller.go#L220-L247

**Comment by jazka** [comment]
@hwdef @william-wang @k82cn I switched to V1.4.0 and there is panic like below, please to help to check what happened.

https://l4x826wg3c.feishu.cn/file/boxcnXsgfyXeRa24rMfm5Co794e
https://l4x826wg3c.feishu.cn/file/boxcnqajf50I2nxyWTfTDW2QQUc
https://l4x826wg3c.feishu.cn/file/boxcnZFDWS4Cp2hUbL5O4lYjlnc

**Comment by william-wang** [comment]
@jazka the issue is fixed in pr #1685, unfortunately it is not merged to release-1.4 branch. I will cherry-pick it 1.4.

**Comment by jazka** [comment]
> @jazka the issue is fixed in pr #1685, unfortunately it is not merged to release-1.4 branch. I will cherry-pick it 1.4.

Do you mean the image of volcanosh/vc-scheduler:v1.4.0 will be updated with this bug fix?

**Comment by william-wang** [comment]
> > @jazka the issue is fixed in pr #1685, unfortunately it is not merged to release-1.4 branch. I will cherry-pick it 1.4.
> 
> Do you mean the image of volcanosh/vc-scheduler:v1.4.0 will be updated with this bug fix?

Yes.  That's right.

**Comment by jazka** [comment]
> > > @jazka the issue is fixed in pr #1685, unfortunately it is not merged to release-1.4 branch. I will cherry-pick it 1.4.
> > 
> > 
> > Do you mean the image of volcanosh/vc-scheduler:v1.4.0 will be updated with this bug fix?
> 
> Yes. That's right.

Cool, when will it be updated？I need to update it in our cluster ASAP and test more.

BTW, is there any progress on the stuck issue? Any cases I need to try?

**Comment by william-wang** [comment]
> If you want to reporduce this issue, it is better to test with training-operator pytorchjob, since it is in my enviroment and it will create podgroup by call volcano sdk directly. Think the case below:
> 
> 1. Disable gang-scheduler in training-operator, then it will create pytorchjob, and without podgroup. The pods of pytorchjob will be scheduled by "default-scheduler"
> 2. Enable gang-scheduler in training-operator, then it will reconcile and call volcano sdk to create podgourp for the running job, even its pods has been running with "schedulerName: default-scheduler"
>    I am not sure if it is related with volcano stuck. I am also tring to find the way to reproduce this issue.
> 
@jazka It's not a general usage. We userally enable the gang-scheduler in operator and use the volcano directly.
May I know why you disable gang-scheduler and use "default scheduler" to schedule the pods of pytorch job and then enable the gang-scheduler with volcano.  This will not take advantage of volcano's capabilities such as gang-scheduling. 
Additionally, I think Volcano would skip the pod with "schedulerName: default-scheduler".

**Comment by jazka** [comment]
@william-wang We did not swith "disabale/enable" on purpose. 

At the beginning, volcano is not used in our cluster and training-operator is used with gang-scheduler disabled. Many pytorch jobs are running on this config. One day, we ran into the problem and need to enable gang-scheduler. Then we deployed volcano and enable gang-scheduler in training-operator. After running for a while with gang-scheduler, the stuck issue is happened which will pend all the job. It will resume after restarting volcano pods, but it still will happen later. We have to disable gang-scheduler in training-operator. 

I just posted this usage since I am not sure whether it is related with the stuck issue.

Review the code, I did not find any possible place to make this stuck issue. Do you have any guess which I need to verify?

**Comment by jazka** [comment]
> > > @jazka the issue is fixed in pr #1685, unfortunately it is not merged to release-1.4 branch. I will cherry-pick it 1.4.
> > 
> > 
> > Do you mean the image of volcanosh/vc-scheduler:v1.4.0 will be updated with this bug fix?
> 
> Yes. That's right.

@william-wang Do you have a plan when the new v1.4.0 image is updated?

**Comment by william-wang** [comment]
@jazka I have submited the fix to release 1.4 branch. however the CI could not pass because the ginkgo version is asked to upgrade by github. It need to take some time to upgrade.
From my point of view, using **1.5-beta** version might be a better choice for you to have quick verification. I just insatlled the 1.5-beta and run the pytorch job. It works as expected.

**Comment by jazka** [comment]
@william-wang I do not find 1.5-beta image from https://hub.docker.com/r/volcanosh/vc-scheduler/tags, where can I pull it?

**Comment by william-wang** [comment]
@jazka use "docker pull  volcanosh/vc-scheduler:v1.5.0"

**Comment by jazka** [comment]
> @jazka use "docker pull volcanosh/vc-scheduler:v1.5.0"

@william-wang Do you suggest we use volcano v1.5.0, not v1.5.0-beta?

**Comment by william-wang** [comment]
> @william-wang Do you suggest we use volcano v1.5.0, not v1.5.0-beta? Is the image "volcanosh/vc-scheduler:v1.5.0" builded on code of git tag v1.5.0-Beta? So I just need to redeploy volcano by "https://raw.githubusercontent.com/volcano-sh/volcano/release-1.5/installer/volcano-development.yaml", right?
Yes, you are right :)

**Comment by jazka** [comment]
@william-wang The stuck issue is still existed with release v1.5.0. Please help to check the logs of volcano-scheduler below. Do you have any way to online chat? I need your help since this issue blocks our product business, thx.

![image](https://user-images.githubusercontent.com/7553111/149548948-bf0aea7b-fa3d-4c12-abe1-9002aa68ca66.png)

**Comment by justadogistaken** [comment]
There is an risk of blocking in "NamespaceCollection.Snapshot" => "n.quotaWeight.Pop()". I think the blocking problem is caused by this place. see the `len(h.data.queue) == 0`
```
func (h *Heap) Pop() (interface{}, error) {
	h.lock.Lock()
	defer h.lock.Unlock()
	for len(h.data.queue) == 0 {
		// When the queue is empty, invocation of Pop() is blocked until new item is enqueued.
		// When Close() is called, the h.closed is set and the condition is broadcast,
		// which causes this loop to continue and return from the Pop().
		if h.closed {
			return nil, fmt.Errorf("heap is closed")
		}
		h.cond.Wait()
	}
	obj := heap.Pop(h.data)
	if obj == nil {
		return nil, fmt.Errorf("object was removed from heap data")
	}

	return obj, nil
}
```

**Comment by justadogistaken** [comment]
There is an risk of blocking in "NamespaceCollection.Snapshot" => "n.quotaWeight.Pop()".  Is there a extreme case that resourceQuata has been deleted while "snapshot"?

```
func (h *Heap) Pop() (interface{}, error) {
	h.lock.Lock()
	defer h.lock.Unlock()
	for len(h.data.queue) == 0 {
		// When the queue is empty, invocation of Pop() is blocked until new item is enqueued.
		// When Close() is called, the h.closed is set and the condition is broadcast,
		// which causes this loop to continue and return from the Pop().
		if h.closed {
			return nil, fmt.Errorf("heap is closed")
		}
		h.cond.Wait()
	}
	obj := heap.Pop(h.data)
	if obj == nil {
		return nil, fmt.Errorf("object was removed from heap data")
	}

	return obj, nil
}
```

**Comment by jazka** [comment]
> There is an risk of blocking in "NamespaceCollection.Snapshot" => "n.quotaWeight.Pop()". Is there a extreme case that resourceQuata has been deleted while "snapshot"?
> 
> ```
> func (h *Heap) Pop() (interface{}, error) {
> 	h.lock.Lock()
> 	defer h.lock.Unlock()
> 	for len(h.data.queue) == 0 {
> 		// When the queue is empty, invocation of Pop() is blocked until new item is enqueued.
> 		// When Close() is called, the h.closed is set and the condition is broadcast,
> 		// which causes this loop to continue and return from the Pop().
> 		if h.closed {
> 			return nil, fmt.Errorf("heap is closed")
> 		}
> 		h.cond.Wait()
> 	}
> 	obj := heap.Pop(h.data)
> 	if obj == nil {
> 		return nil, fmt.Errorf("object was removed from heap data")
> 	}
> 
> 	return obj, nil
> }
> ```

Yes, you are right. I reproduced this stuck issue when deleting resourceQuota. @william-wang will try to reproduced it and fix it.

**Comment by shinytang6** [comment]
> There is an risk of blocking in "NamespaceCollection.Snapshot" => "n.quotaWeight.Pop()". Is there a extreme case that resourceQuata has been deleted while "snapshot"?
> 
> ```
> func (h *Heap) Pop() (interface{}, error) {
> 	h.lock.Lock()
> 	defer h.lock.Unlock()
> 	for len(h.data.queue) == 0 {
> 		// When the queue is empty, invocation of Pop() is blocked until new item is enqueued.
> 		// When Close() is called, the h.closed is set and the condition is broadcast,
> 		// which causes this loop to continue and return from the Pop().
> 		if h.closed {
> 			return nil, fmt.Errorf("heap is closed")
> 		}
> 		h.cond.Wait()
> 	}
> 	obj := heap.Pop(h.data)
> 	if obj == nil {
> 		return nil, fmt.Errorf("object was removed from heap data")
> 	}
> 
> 	return obj, nil
> }
> ```

good catch! It seems that `NamespaceCollection` is always growing.

**Comment by justadogistaken** [comment]
I recommed we use priorityQueue execluding "sync.Cond". Snapshot does not need to wait for producer. Just get a copy of current context. 
I can do something to help if we reach an agreement on replacing the priorityQueue tool.

**Comment by shinytang6** [comment]
l agree that using PriorityQueue here.
/cc @william-wang

**Comment by william-wang** [comment]
@jazka 
Reproduce step:
1. Create namespace `ns-1`
2. Create resourequota for `ns-1`
[root@volcano-1951-39604 create]# kubectl apply -f quota-1.yaml --namespace=ns-1
resourcequota/quota-1 created
3. The scheduling works well
4. Delete all resourcequota for ns-1
[root@volcano-1951-39604 create]# kubectl delete resourcequota quota-1 -n ns-1
resourcequota "quota-1" deleted
5. Volcnao scheduler pod gets stuck
I0118 15:19:21.069962       8 session.go:192] Close Session 39e3a9e5-a2c6-4323-b808-4edb50b17a54
I0118 15:19:21.069969       8 scheduler.go:111] End scheduling ...
I0118 15:19:21.503880       8 event_handlers.go:741] Delete ResourceQuota <ns-1/quota-1> in cache
I0118 15:19:22.070077       8 scheduler.go:92] Start scheduling ...
I0118 15:19:22.070143       8 node_info.go:274] set the node 10.1.17.224 status to Ready.
I0118 15:19:22.070187       8 cache.go:953] Namespace kube-system has weight 1
I0118 15:19:22.070202       8 cache.go:953] Namespace kubeflow has weight 1
...
6. Submit new workload and they can not be scheduled.
7. The scheduler pod works well again after a restart.

The casue is as below:
The `Quota service` in your environment delete all `resourcequota` resource from time to time.
Whenever there is a namespce kept in sc.NamespaceCollection and all the conresponding resourcequota are deleted. the below snapshot function would block at Pop() in client-go.
```
func (n *NamespaceCollection) Snapshot() *NamespaceInfo {
	var weight int64 = DefaultNamespaceWeight

	obj, err := n.quotaWeight.Pop()
```
The Volcano 0.2 version had considered this problem and provided solution in namespace fairshare feature.
Volcano always keeps a default quota item in the heap even when there is no resourequota. The Pop() would never blocks.
Unfortunately the feature was broken in the later code changes.  Please have a try with the PR#1971.

**Comment by jazka** [comment]
Hi, the PR#1971 has been merged master branch, will you cherry-pick it to release 1.5.0? Or which docker image on dockerhub I should use?

**Comment by Thor-wl** [comment]
> Hi, the PR#1971 has been merged master branch, will you cherry-pick it to release 1.5.0? Or which docker image on dockerhub I should use?

Yes, this PR will be merged into v1.5.0 which is planed to be released around the late January. Then we will also push a docker image with tag v1.5.0 to the dockerhub.

**Comment by jazka** [comment]
> > Hi, the PR#1971 has been merged master branch, will you cherry-pick it to release 1.5.0? Or which docker image on dockerhub I should use?
> 
> Yes, this PR will be merged into v1.5.0 which is planed to be released around the late January. Then we will also push a docker image with tag v1.5.0 to the dockerhub.

@Thor-wl The release-1.5 has merged this bug fix, but the image with tag v1.5.0 in dockerhub is still old, do you have any plan when to publish new image? Or should I build the image myself from release-1.5?

**Comment by Thor-wl** [comment]
> 

That's strange. Let me take a check. Yes, we plan to release the v1.5.0 recently and then will push a new image to dockerhub. If it is urgent for you, you can also build an image yourself from branch release-1.5.

**Comment by william-wang** [comment]
@Thor-wl Can we make the ci push the new image to dockerhub once a new pr is merged?

**Comment by Thor-wl** [comment]
> @Thor-wl Can we make the ci push the new image to dockerhub once a new pr is merged?

We have made a periodic task in CI to build the latest image and push it to the dockerhub everyday. I think that's also available.
